File: 57.insert-interval.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> insert(vector<Interval>& intervals, Interval newInterval) {
        vector<Interval> ans;
        auto it = intervals.begin();
        for (; it < intervals.end(); it++) {
            int start = it->start;
            int end = it->end;
            if (newInterval.start > end) {
                ans.push_back(*it);
            } else if (newInterval.end < start) {
                break;
            } else {
                newInterval.start = min(start, newInterval.start);
                newInterval.end = max(end, newInterval.end);
            }
        }
        ans.push_back(newInterval);
        ans.insert(ans.end(), it, intervals.end());
        return ans;
    }
};

File: 57.insert-interval.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    def insert(self, intervals, newInterval):
        """
        :type intervals: List[Interval]
        :type newInterval: Interval
        :rtype: List[Interval]
        """
        ans = []
        i = 0
        while i < len(intervals):
            start = intervals[i].start
            end = intervals[i].end
            if newInterval.start > end:
                ans.append(intervals[i])
            elif newInterval.end < start:
                break
            else:
                newInterval.start = min(start, newInterval.start)
                newInterval.end = max(end, newInterval.end)
            i += 1
        ans.append(newInterval)
        ans += intervals[i:]
        return ans